test/core/parachain/grid_tracker.cpp

```diff
@@ -0,0 +1,900 @@
+                               .statement_knowledge = create_filter(
+                                   {{false, true, false}, {true, false, true}}),
+                           },
+                           ManifestKind::Full,
+                           ValidatorIndex(1))
+          .error(),
+      GridTracker::Error::DISALLOWED_DIRECTION);
+
+  // Unknown group
+  ASSERT_EQ(
+      tracker
+          .import_manifest(session_topology,
+                           groups,
+                           candidate_hash,
+                           3,
+                           ManifestSummary{
+                               .claimed_parent_hash = fromNumber(0),
+                               .claimed_group_index = GroupIndex(1),
+                               .statement_knowledge = create_filter(
+                                   {{false, true, false}, {true, false, true}}),
+                           },
+                           ManifestKind::Full,
+                           ValidatorIndex(0))
+          .error(),
+      GridTracker::Error::DISALLOWED_GROUP_INDEX);
+}
+
+TEST_F(GridTrackerTest, reject_malformed_wrong_group_size) {
+  GridTracker tracker;
+
+  SessionTopologyView session_topology = {View{
+      .sending = {},
+      .receiving = {ValidatorIndex(0)},
+  }};
+
+  const auto groups = dummy_groups(3);
+
+  const auto candidate_hash = fromNumber(42);
+  ASSERT_EQ(groups.get_size_and_backing_threshold(GroupIndex(0)),
+            std::make_tuple(size_t(3), size_t(2)));
+
+  ASSERT_EQ(tracker
+                .import_manifest(
+                    session_topology,
+                    groups,
+                    candidate_hash,
+                    3,
+                    ManifestSummary{
+                        .claimed_parent_hash = fromNumber(0),
+                        .claimed_group_index = GroupIndex(0),
+                        .statement_knowledge = create_filter(
+                            {{false, true, false, true}, {true, false, true}}),
+                    },
+                    ManifestKind::Full,
+                    ValidatorIndex(0))
+                .error(),
+            GridTracker::Error::MALFORMED_REMOTE_KNOWLEDGE_LEN);
+
+  ASSERT_EQ(tracker
+                .import_manifest(
+                    session_topology,
+                    groups,
+                    candidate_hash,
+                    3,
+                    ManifestSummary{
+                        .claimed_parent_hash = fromNumber(0),
+                        .claimed_group_index = GroupIndex(0),
+                        .statement_knowledge = create_filter(
+                            {{false, true, false}, {true, false, true, false}}),
+                    },
+                    ManifestKind::Full,
+                    ValidatorIndex(0))
+                .error(),
+            GridTracker::Error::MALFORMED_REMOTE_KNOWLEDGE_LEN);
+}
+
+TEST_F(GridTrackerTest, reject_malformed_no_seconders) {
```

Comment:
Missing given-when-then test notation but it seems we gave up on this a long time ago.